### PR TITLE
refactor integrated alerts algo and add dist alerts

### PR DESCRIPTION
**Pull request by solomon-negusse**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the new behavior?

Adds algorithm to decode and render DIST alerts

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: app/routes/titiler/algorithms/alerts.py [hunk 1]
```diff
@@ -9,31 +9,29 @@
...
+            DeforestationAlertConfidence.high: {
+                "confidence": 3,
+                "colors": (220, 102, 153),
+            },
+            DeforestationAlertConfidence.highest: {
+                "confidence": 4,
+                "colors": (201, 42, 109),
+            },
+        }
+    )
```
Comment (gtempus) [inline app/routes/titiler/algorithms/alerts.py]: I ❤️ going with the immutable tuple here, @solomon-negusse.

I also like the descriptiveness of the _red, green, blue_ keys of the original `dict`. 

Did you consider using a namedTuple to get the best of both worlds?
Something like:
```python
# Define a namedtuple for Colors (immutable)
Colors = namedtuple('Colors', ['red', 'green', 'blue'])

# Define a namedtuple for DeforestationAlert (immutable)
DeforestationAlert = namedtuple('DeforestationAlert', ['confidence', 'colors'])

# Assuming DeforestationAlertConfidence is an Enum or similar object
conf_colors: OrderedDict = OrderedDict(
    {
        DeforestationAlertConfidence.low: DeforestationAlert(confidence=2, colors=Colors(237, 164, 194)),
        DeforestationAlertConfidence.high: DeforestationAlert(confidence=3, colors=Colors(220, 102, 153)),
        DeforestationAlertConfidence.highest: DeforestationAlert(confidence=4, colors=Colors(201, 42, 109)),
    }
)
```

File: app/routes/titiler/algorithms/alerts.py [hunk 2]
```diff
@@ -9,31 +9,29 @@
...
+    record_start_date: str = "2014-12-31"
 
     # Parameters
     default_start_date: str = (date.today() - relativedelta(days=180)).strftime(
```
Comment (gtempus) [inline app/routes/titiler/algorithms/alerts.py]: This is a nitpick but I thought I'd mention it:
I noticed that `data.today()` is called a couple of times. Is it worth it to just call it once? 🤷

**Review by gtempus — COMMENTED** [review]
I think you made a good design decision, considering there are only two types of `Alerts` in our TiTiler `Algorithm` domain right now. 

Regarding `__call__` and its current structure:
It's currently in a long imperative style. What about breaking it up into its principal parts? 

Namely: 
- assigning RGB values
- creating the mask
- and applying the mask. 

It would give it a more declarative feel, and tests could zero in on specific behavior at the unit level. 

I was going to try to write a few tests myself, but I ran out of time. 😦

File: app/routes/titiler/algorithms/alerts.py (hunk 1 above)
Comment (solomon-negusse) [reply app/routes/titiler/algorithms/alerts.py]: Great idea, thanks!

File: app/routes/titiler/algorithms/alerts.py (hunk 2 above)
Comment (solomon-negusse) [reply app/routes/titiler/algorithms/alerts.py]: Always good to DRY,  let me update it.

**Comment by solomon-negusse** [comment]
> I think you made a good design decision, considering there are only two types of `Alerts` in our TiTiler `Algorithm` domain right now.
> 
> Regarding `__call__` and its current structure: It's currently in a long imperative style. What about breaking it up into its principal parts?
> 
> Namely:
> 
>     * assigning RGB values
> 
>     * creating the mask
> 
>     * and applying the mask.
> 
> 
> It would give it a more declarative feel, and tests could zero in on specific behavior at the unit level.
> 
> I was going to try to write a few tests myself, but I ran out of time. 😦

Thanks for the feedback, @gtempus.  I have refactored the `Alerts` class and added tests.

File: app/routes/titiler/algorithms/dist_alerts.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            DeforestationAlertConfidence.low: DeforestationAlert(
+                confidence=2, colors=Colors(237, 164, 194)
+            ),
+            DeforestationAlertConfidence.high: DeforestationAlert(
+                confidence=3, colors=Colors(220, 102, 153)
+            ),
+        }
+    )
+
+    record_start_date: str = "2020-12-13"
```
Comment (danscales) [inline app/routes/titiler/algorithms/dist_alerts.py]: Did you flip the digits?  The start date is "2020-12-31", according to the spec on VEG-DIST-DATE

File: app/routes/titiler/algorithms/dist_alerts.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+from collections import OrderedDict, namedtuple
+from .alerts import Alerts
+
+from app.models.enumerators.alerts_confidence import DeforestationAlertConfidence
+
+
+class DISTAlerts(Alerts):
+    title: str = "Land Disturbunce (DIST) Alerts"
```
Comment (danscales) [inline app/routes/titiler/algorithms/dist_alerts.py]: Disturbunce -> Disturbance

File: app/routes/titiler/algorithms/dist_alerts.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    description: str = "Decode and visualize DIST alerts"
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
+    DeforestationAlert: namedtuple = namedtuple(
+        "DeforestationAlert", ["confidence", "colors"]
```
Comment (danscales) [inline app/routes/titiler/algorithms/dist_alerts.py]: Should this be "DisturbanceAlert" ?  DIST-Alerts covers more than forests?

File: app/routes/titiler/algorithms/dist_alerts.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+from .alerts import Alerts
+
+from app.models.enumerators.alerts_confidence import DeforestationAlertConfidence
+
+
+class DISTAlerts(Alerts):
+    title: str = "Land Disturbunce (DIST) Alerts"
+    description: str = "Decode and visualize DIST alerts"
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
```
Comment (danscales) [inline app/routes/titiler/algorithms/dist_alerts.py]: Any need to include this if it is the same as the Base Class?

File: app/routes/titiler/algorithms/integrated_alerts.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+from .alerts import Alerts
+
+from app.models.enumerators.alerts_confidence import DeforestationAlertConfidence
+
+
+class IntegratedAlerts(Alerts):
+    title: str = "Integrated Deforestation Alerts"
+    description: str = "Decode and visualize deforestation alerts"
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
```
Comment (danscales) [inline app/routes/titiler/algorithms/integrated_alerts.py]: Same here, any reason to include Colors here if it's the same as the base class?

**Review by danscales — APPROVED** [review]

File: app/routes/titiler/algorithms/dist_alerts.py (hunk 3 above)
Comment (solomon-negusse) [reply app/routes/titiler/algorithms/dist_alerts.py]: Ah, typo in our confluence document where I took this from then.

File: app/routes/titiler/algorithms/dist_alerts.py (hunk 6 above)
Comment (solomon-negusse) [reply app/routes/titiler/algorithms/dist_alerts.py]: Yes, they are not accessible to the subclasses. I ended up taking them out of the class and reusing them both in dist and integrated alerts modules now.

File: app/routes/titiler/algorithms/dist_alerts.py (hunk 5 above)
Comment (solomon-negusse) [reply app/routes/titiler/algorithms/dist_alerts.py]: Since there's just a single definition of that in alerts.py now, I ended up calling it `AlertConfig` and also removed `deforestation` from a couple of variables that are common to both alert systems.

**Review by gtempus — APPROVED** [review]
This looks great @solomon-negusse! 💪 
It's well organized, there is a separation of concerns, and it's tested. 💥
